src/Compilers/Core/Portable/Syntax/GreenNode.cs

```diff
@@ -566,36 +621,116 @@ internal DiagnosticInfo[] GetDiagnostics()
         #endregion
 
         #region Text
-        public abstract string ToFullString();
 
-        public virtual void WriteTo(System.IO.TextWriter writer)
+        public virtual string ToFullString()
         {
-            this.WriteTo(writer, true, true);
+            var sb = PooledStringBuilder.GetInstance();
+            var writer = new System.IO.StringWriter(sb.Builder, System.Globalization.CultureInfo.InvariantCulture);
+            this.WriteTo(writer, leading: true, trailing: true);
+            return sb.ToStringAndFree();
         }
 
-        protected internal virtual void WriteTo(System.IO.TextWriter writer, bool leading, bool trailing)
+        public override string ToString()
         {
-            bool first = true;
-            int n = this.SlotCount;
-            int lastIndex = n - 1;
-            for (; lastIndex >= 0; lastIndex--)
+            var sb = PooledStringBuilder.GetInstance();
+            var writer = new System.IO.StringWriter(sb.Builder, System.Globalization.CultureInfo.InvariantCulture);
+            this.WriteTo(writer, leading: false, trailing: false);
+            return sb.ToStringAndFree();
+        }
+
+        public void WriteTo(System.IO.TextWriter writer)
+        {
+            this.WriteTo(writer, leading: true, trailing: true);
+        }
+
+        protected internal void WriteTo(TextWriter writer, bool leading, bool trailing)
+        {
+            // Use an actual Stack so we can write out deeply recursive structures without overflowing.
+            var stack = new Stack<ValueTuple<GreenNode, bool, bool>>();
+            stack.Push(ValueTuple.Create(this, leading, trailing));
+
+            // Separated out stack processing logic so that it does not unintentially refer to 
+            // "this", "leading" or "trailing.
+            ProcessStack(writer, stack);
+        }
+
+        private static void ProcessStack(TextWriter writer, Stack<ValueTuple<GreenNode, bool, bool>> stack)
+        {
+            while (stack.Count > 0)
             {
-                var child = this.GetSlot(lastIndex);
+                var current = stack.Pop();
+                var currentNode = current.Item1;
+                var currentLeading = current.Item2;
+                var currentTrailing = current.Item3;
+
+                if (currentNode.IsToken)
+                {
+                    currentNode.WriteTokenTo(writer, currentLeading, currentTrailing);
+                    continue;
+                }
+
+                if (currentNode.IsTrivia)
+                {
+                    currentNode.WriteTriviaTo(writer);
+                    continue;
+                }
+
+                var firstIndex = GetFirstNonNullChildIndex(currentNode);
+                var lastIndex = GetLastNonNullChildIndex(currentNode);
```

Comment:
It looks like `GetFirst/LastNonNullChildIndex` are similar to the loop below. Is this the only place we're using those methods? If so, consider removing those methods and looping over all items here instead.